tests/generation/test_utils.py

```diff
@@ -3493,3 +3494,77 @@ def test_generate_from_inputs_embeds_with_bos_token_id_is_none(self):
         # bos_token_id is required when no input ids nor inputs_embeds is passed
         with self.assertRaises(ValueError):
             model.generate(max_length=20, bos_token_id=None)
+
+
+class TestGetTokensDiag(unittest.TestCase):
+    def test_no_intersection(self):
+        prompt = np.array([[1, 2, 3]])
+        prompt_plus_new_tokens = np.array([[4, 5, 6]])
+        result = AssistedCandidateGeneratorDifferentTokenizers._get_tokens_diag(prompt, prompt_plus_new_tokens)
+        self.assertEqual(result, (None, None, None, None, None))
+
+    def test_complete_overlap(self):
+        prompt = np.array([[1, 2, 3]])
+        prompt_plus_new_tokens = np.array([[1, 2, 3, 4, 5]])
+        replace_tokens_from_prompt, disrep_length, new_tokens_with_disrep, new_tokens_only, discrep_only = (
+            AssistedCandidateGeneratorDifferentTokenizers._get_tokens_diag(prompt, prompt_plus_new_tokens)
+        )
+        np.testing.assert_array_equal(replace_tokens_from_prompt, np.array([[]]))
+        self.assertEqual(disrep_length, 0)
+        np.testing.assert_array_equal(new_tokens_with_disrep, np.array([[4, 5]]))
+        np.testing.assert_array_equal(new_tokens_only, np.array([[4, 5]]))
+        np.testing.assert_array_equal(discrep_only, np.array([[]]))
+
+    def test_partial_overlap(self):
+        prompt = np.array([[1, 2, 3]])
+        prompt_plus_new_tokens = np.array([[2, 3, 4, 5]])
+        replace_tokens_from_prompt, disrep_length, new_tokens_with_disrep, new_tokens_only, discrep_only = (
+            AssistedCandidateGeneratorDifferentTokenizers._get_tokens_diag(prompt, prompt_plus_new_tokens)
+        )
+        np.testing.assert_array_equal(replace_tokens_from_prompt, np.array([[]]))
+        self.assertEqual(disrep_length, 0)
+        np.testing.assert_array_equal(new_tokens_with_disrep, np.array([[4, 5]]))
+        np.testing.assert_array_equal(new_tokens_only, np.array([[4, 5]]))
+        np.testing.assert_array_equal(discrep_only, np.array([[]]))
+
+    def test_no_new_tokens(self):
+        prompt = np.array([[1, 2, 3]])
+        prompt_plus_new_tokens = np.array([[1, 2, 3]])
+        replace_tokens_from_prompt, disrep_length, new_tokens_with_disrep, new_tokens_only, discrep_only = (
+            AssistedCandidateGeneratorDifferentTokenizers._get_tokens_diag(prompt, prompt_plus_new_tokens)
+        )
+        np.testing.assert_array_equal(replace_tokens_from_prompt, np.array([[]]))
+        self.assertEqual(disrep_length, 0)
+        np.testing.assert_array_equal(new_tokens_with_disrep, np.array([[]]))
+        np.testing.assert_array_equal(new_tokens_only, np.array([[]]))
+        np.testing.assert_array_equal(discrep_only, np.array([[]]))
+
+
+class TestAnyTokenizerGenerationCorrectness(unittest.TestCase):
+    def test_speculative_decoding_equals_regular_decoding(self):
```

Comment:
Nit: let's move this to the integration tests above (`GenerationIntegrationTests`). We place all `generate` integration tests there 🤗